diff -Naur xen/console/console.c xen.patched/console/console.c
--- xen/console/console.c	2005-09-23 10:54:50.000000000 -0600
+++ xen.patched/console/console.c	2005-09-23 10:57:51.000000000 -0600
@@ -768,9 +768,16 @@
 #endif
 
 	if (xen_start_info->flags & SIF_INITDOMAIN) {
+#ifdef __ia64__
+		xencons_priv_irq = bind_virq_to_evtchn(VIRQ_CONSOLE);
+		bind_evtchn_to_irqhandler(xencons_priv_irq,
+				xencons_priv_interrupt, 0, "console", NULL);
+#else
+
 		xencons_priv_irq = bind_virq_to_irq(VIRQ_CONSOLE);
 		(void)request_irq(xencons_priv_irq,
 				  xencons_priv_interrupt, 0, "console", NULL);
+#endif
 	} else {
 		xencons_ring_register_receiver(xencons_rx);
 	}
