diff -Naur xen/core/devmem.c xen.patched/core/devmem.c
--- xen/core/devmem.c	2005-09-23 10:54:50.000000000 -0600
+++ xen.patched/core/devmem.c	2005-09-23 10:57:51.000000000 -0600
@@ -93,6 +93,9 @@
 	if (uncached_access(file))
 		vma->vm_page_prot = pgprot_noncached(vma->vm_page_prot);
 
+#ifdef __ia64__
+#define	direct_remap_pfn_range(a,b,c,d,e,f)	remap_pfn_range(a,b,c,d,e)
+#endif
 	if (direct_remap_pfn_range(vma, vma->vm_start, vma->vm_pgoff,
 				   vma->vm_end - vma->vm_start,
 				   vma->vm_page_prot, DOMID_IO))
