diff -Naur xen/core/gnttab.c xen.patched/core/gnttab.c
--- xen/core/gnttab.c	2005-09-23 10:54:50.000000000 -0600
+++ xen.patched/core/gnttab.c	2005-09-23 10:57:51.000000000 -0600
@@ -346,6 +350,10 @@
 	if ( hypercall.op != __HYPERVISOR_grant_table_op )
 		return -ENOSYS;
 
+
+#ifdef __ia64__
+	ret = HYPERVISOR_grant_table_op(hypercall.arg[0], (void *)hypercall.arg[1], hypercall.arg[2]);
+#else
 	/* hypercall-invoking asm taken from privcmd.c */
 	__asm__ __volatile__ (
 		"pushl %%ebx; pushl %%ecx; pushl %%edx; "
@@ -359,6 +367,7 @@
 		TRAP_INSTR "; "
 		"popl %%edi; popl %%esi; popl %%edx; popl %%ecx; popl %%ebx"
 		: "=a" (ret) : "0" (&hypercall) : "memory" );
+#endif
 
 	return ret;
 }
@@ -423,8 +432,13 @@
 	BUG_ON(HYPERVISOR_grant_table_op(GNTTABOP_setup_table, &setup, 1));
 	BUG_ON(setup.status != 0);
 
+#ifdef __ia64__
+	shared = __va(frames[0] << PAGE_SHIFT);
+	printk("grant table at %p\n", shared);
+#else
 	for (i = 0; i < NR_GRANT_FRAMES; i++)
 		set_fixmap(FIX_GNTTAB_END - i, frames[i] << PAGE_SHIFT);
+#endif
 
 	return 0;
 }
@@ -450,7 +466,9 @@
 
 	BUG_ON(gnttab_resume());
 
+#ifndef __ia64__
 	shared = (grant_entry_t *)fix_to_virt(FIX_GNTTAB_END);
+#endif
 
 	for (i = NR_RESERVED_ENTRIES; i < NR_GRANT_ENTRIES; i++)
 		gnttab_list[i] = i + 1;
