diff -Naur xen/privcmd/privcmd.c xen.patched/privcmd/privcmd.c
--- xen/privcmd/privcmd.c	2005-09-23 10:54:50.000000000 -0600
+++ xen.patched/privcmd/privcmd.c	2005-09-23 10:57:51.000000000 -0600
@@ -180,6 +183,15 @@
 		for (i = 0; i < m.num; i++, addr += PAGE_SIZE, p++) {
 			if (get_user(mfn, p))
 				return -EFAULT;
+#ifdef __ia64__
+			ret = remap_pfn_range(vma,
+					      addr&PAGE_MASK,
+					      mfn,
+					      1<<PAGE_SHIFT,
+					      vma->vm_page_prot);
+			if (ret < 0)
+			    goto batch_err;
+#else
 
 			ret = create_lookup_pte_addr(vma->vm_mm, addr, &ptep);
 			if (ret)
@@ -190,6 +202,7 @@
 
 			if (HYPERVISOR_mmu_update(&u, 1, NULL, m.dom) < 0)
 				put_user(0xF0000000 | mfn, p);
+#endif
 		}
 
 		ret = 0;
@@ -205,6 +218,7 @@
 	break;
 #endif
 
+#ifndef __ia64__
 	case IOCTL_PRIVCMD_GET_MACH2PHYS_START_MFN: {
 		unsigned long m2pv = (unsigned long)machine_to_phys_mapping;
 		pgd_t *pgd = pgd_offset_k(m2pv);
@@ -216,6 +230,7 @@
 			-EFAULT: 0;
 	}
 	break;
+#endif
 
 	default:
 		ret = -EINVAL;
