diff -u linux-2.6.12/kernel/ptrace.c linux-2.6.12-pdb/kernel/ptrace.c
--- linux-2.6.12/kernel/ptrace.c       2005-06-17 20:48:29.000000000 +0100
+++ linux-2.6.12-pdb/kernel/ptrace.c   2005-07-22 13:23:16.000000000 +0100
@@ -239,6 +239,7 @@
 
        return buf - old_buf;
 }
+EXPORT_SYMBOL(access_process_vm);
 
 int ptrace_readdata(struct task_struct *tsk, unsigned long src, char __user *dst, int len)
 {
