508c578140642a641bb9b888369719c510ae2a00
diff --git a/net/core/dev.c b/net/core/dev.c
index e814a89..240773b 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -1087,11 +1087,6 @@ int skb_checksum_help(struct sk_buff *sk
 		goto out_set_summed;
 
 	if (unlikely(skb_shinfo(skb)->gso_size)) {
-		static int warned;
-
-		WARN_ON(!warned);
-		warned = 1;
-
 		/* Let GSO fix up the checksum. */
 		goto out_set_summed;
 	}
@@ -1141,11 +1136,6 @@ struct sk_buff *skb_gso_segment(struct s
 	__skb_pull(skb, skb->mac_len);
 
 	if (unlikely(skb->ip_summed != CHECKSUM_HW)) {
-		static int warned;
-
-		WARN_ON(!warned);
-		warned = 1;
-
 		if (skb_header_cloned(skb) &&
 		    (err = pskb_expand_head(skb, 0, 0, GFP_ATOMIC)))
 			return ERR_PTR(err);
