diff --git a/src/arch/i386/prefix/romprefix.S b/src/arch/i386/prefix/romprefix.S
index 0f92415..cce7505 100644
--- a/src/arch/i386/prefix/romprefix.S
+++ b/src/arch/i386/prefix/romprefix.S
@@ -391,6 +391,7 @@ no_pmm:
 	xorw	%di, %di
 	cs rep	movsb
 
+#ifndef NO_POST_PROMPT
 	/* Prompt for POST-time shell */
 	movw	$init_message_prompt, %si
 	xorw	%di, %di
@@ -418,6 +419,7 @@ no_pmm:
 	pushw	%cs
 	call	exec
 2:
+#endif
 	/* Restore registers */
 	popw	%gs
 	popw	%fs
@@ -546,6 +548,7 @@ init_message_pmm:
 init_message_int19:
 	.asciz	" INT19"
 	.size	init_message_int19, . - init_message_int19
+#ifndef NO_POST_PROMPT
 init_message_prompt:
 	.asciz	"\nPress Ctrl-B to configure "
 	.size	init_message_prompt, . - init_message_prompt
@@ -555,6 +558,7 @@ init_message_dots:
 init_message_done:
 	.asciz	"\n\n"
 	.size	init_message_done, . - init_message_done
+#endif
 
 /* PCI bus:dev.fn
  *
