diff -pruN gpxe/src/arch/i386/prefix/romprefix.S gpxe.new/src/arch/i386/prefix/romprefix.S
--- gpxe/src/arch/i386/prefix/romprefix.S	2010-06-29 20:31:33.000000000 +0100
+++ gpxe.new/src/arch/i386/prefix/romprefix.S	2010-07-20 10:40:20.000000000 +0100
@@ -458,6 +458,7 @@ no_pmm:
 	xorw	%di, %di
 	cs rep	movsb
 
+#ifndef NO_POST_PROMPT
 	/* Prompt for POST-time shell */
 	movw	$init_message_prompt, %si
 	xorw	%di, %di
@@ -484,6 +485,7 @@ no_pmm:
 	pushw	%cs
 	call	exec
 out:
+#endif
 	/* Restore registers */
 	popw	%gs
 	popw	%fs
@@ -538,6 +540,7 @@ init_message_no_pmm:
 init_message_int19:
 	.asciz	" INT19"
 	.size	init_message_int19, . - init_message_int19
+#ifndef NO_POST_PROMPT
 init_message_prompt:
 	.asciz	"\nPress Ctrl-B to configure "
 	.size	init_message_prompt, . - init_message_prompt
@@ -547,6 +550,7 @@ init_message_dots:
 init_message_done:
 	.asciz	"\n\n"
 	.size	init_message_done, . - init_message_done
+#endif
 
 /* ROM image location
  *
