diff -ur linux-2.6.9/drivers/char/drm/ati_pcigart.h linux-2.6.9-new/drivers/char/drm/ati_pcigart.h
--- linux-2.6.9/drivers/char/drm/ati_pcigart.h	2004-10-18 22:55:07.000000000 +0100
+++ linux-2.6.9-new/drivers/char/drm/ati_pcigart.h	2004-11-28 19:42:41.000000000 +0000
@@ -158,7 +158,7 @@
 	ret = 1;
 
 #if defined(__i386__) || defined(__x86_64__)
-	asm volatile ( "wbinvd" ::: "memory" );
+	wbinvd();
 #else
 	mb();
 #endif
