diff -ru linux-2.6.8.1/drivers/mtd/maps/nettel.c linux-2.6.8.1-xen0/drivers/mtd/maps/nettel.c
--- linux-2.6.8.1/drivers/mtd/maps/nettel.c	2004-10-21 15:59:29.000000000 -0600
+++ linux-2.6.8.1-xen0/drivers/mtd/maps/nettel.c	2004-10-22 09:12:17.000000000 -0600
@@ -270,7 +270,7 @@
 	maxsize = AMD_WINDOW_MAXSIZE;
 
 	*amdpar = SC520_PAR(SC520_PAR_BOOTCS, amdaddr, maxsize);
-	__asm__ ("wbinvd");
+	wbinvd();
 
 	nettel_amd_map.phys = amdaddr;
 	nettel_amd_map.virt = (unsigned long)
@@ -382,7 +382,7 @@
 	 */
 	intel1addr = intel0addr + intel0size;
 	*intel1par = SC520_PAR(intel1cs, intel1addr, maxsize);
-	__asm__ ("wbinvd");
+	wbinvd();
 
 	maxsize += intel0size;
 
@@ -408,7 +408,7 @@
 	intel1size = intel_mtd->size - intel0size;
 	if (intel1size > 0) {
 		*intel1par = SC520_PAR(intel1cs, intel1addr, intel1size);
-		__asm__ ("wbinvd");
+		wbinvd();
 	} else {
 		*intel1par = 0;
 	}
