diff -pruN ../orig-linux-2.6.17/arch/i386/kernel/machine_kexec.c ./arch/i386/kernel/machine_kexec.c
--- ../orig-linux-2.6.17/arch/i386/kernel/machine_kexec.c	2006-06-18 02:49:35.000000000 +0100
+++ ./arch/i386/kernel/machine_kexec.c	2007-01-08 15:01:25.000000000 +0000
@@ -133,9 +133,9 @@ typedef asmlinkage NORET_TYPE void (*rel
 					unsigned long start_address,
 					unsigned int has_pae) ATTRIB_NORET;
 
-const extern unsigned char relocate_new_kernel[];
+extern const unsigned char relocate_new_kernel[];
 extern void relocate_new_kernel_end(void);
-const extern unsigned int relocate_new_kernel_size;
+extern const unsigned int relocate_new_kernel_size;
 
 /*
  * A architecture hook called to validate the
diff -pruN ../orig-linux-2.6.17/arch/powerpc/kernel/machine_kexec_32.c ./arch/powerpc/kernel/machine_kexec_32.c
--- ../orig-linux-2.6.17/arch/powerpc/kernel/machine_kexec_32.c	2006-06-18 02:49:35.000000000 +0100
+++ ./arch/powerpc/kernel/machine_kexec_32.c	2007-01-08 15:01:25.000000000 +0000
@@ -30,8 +30,8 @@ typedef NORET_TYPE void (*relocate_new_k
  */
 void default_machine_kexec(struct kimage *image)
 {
-	const extern unsigned char relocate_new_kernel[];
-	const extern unsigned int relocate_new_kernel_size;
+	extern const unsigned char relocate_new_kernel[];
+	extern const unsigned int relocate_new_kernel_size;
 	unsigned long page_list;
 	unsigned long reboot_code_buffer, reboot_code_buffer_phys;
 	relocate_new_kernel_t rnk;
diff -pruN ../orig-linux-2.6.17/arch/ppc/kernel/machine_kexec.c ./arch/ppc/kernel/machine_kexec.c
--- ../orig-linux-2.6.17/arch/ppc/kernel/machine_kexec.c	2006-06-18 02:49:35.000000000 +0100
+++ ./arch/ppc/kernel/machine_kexec.c	2007-01-08 15:01:25.000000000 +0000
@@ -25,8 +25,8 @@ typedef NORET_TYPE void (*relocate_new_k
 				unsigned long reboot_code_buffer,
 				unsigned long start_address) ATTRIB_NORET;
 
-const extern unsigned char relocate_new_kernel[];
-const extern unsigned int relocate_new_kernel_size;
+extern const unsigned char relocate_new_kernel[];
+extern const unsigned int relocate_new_kernel_size;
 
 void machine_shutdown(void)
 {
diff -pruN ../orig-linux-2.6.17/arch/s390/kernel/machine_kexec.c ./arch/s390/kernel/machine_kexec.c
--- ../orig-linux-2.6.17/arch/s390/kernel/machine_kexec.c	2006-06-18 02:49:35.000000000 +0100
+++ ./arch/s390/kernel/machine_kexec.c	2007-01-08 15:01:25.000000000 +0000
@@ -27,8 +27,8 @@ static void kexec_halt_all_cpus(void *);
 
 typedef void (*relocate_kernel_t) (kimage_entry_t *, unsigned long);
 
-const extern unsigned char relocate_kernel[];
-const extern unsigned long long relocate_kernel_len;
+extern const unsigned char relocate_kernel[];
+extern const unsigned long long relocate_kernel_len;
 
 int
 machine_kexec_prepare(struct kimage *image)
diff -pruN ../orig-linux-2.6.17/arch/sh/kernel/machine_kexec.c ./arch/sh/kernel/machine_kexec.c
--- ../orig-linux-2.6.17/arch/sh/kernel/machine_kexec.c	2006-06-18 02:49:35.000000000 +0100
+++ ./arch/sh/kernel/machine_kexec.c	2007-01-08 15:01:25.000000000 +0000
@@ -25,8 +25,8 @@ typedef NORET_TYPE void (*relocate_new_k
 				unsigned long start_address,
 				unsigned long vbr_reg) ATTRIB_NORET;
 
-const extern unsigned char relocate_new_kernel[];
-const extern unsigned int relocate_new_kernel_size;
+extern const unsigned char relocate_new_kernel[];
+extern const unsigned int relocate_new_kernel_size;
 extern void *gdb_vbr_vector;
 
 /*
diff -pruN ../orig-linux-2.6.17/arch/x86_64/kernel/machine_kexec.c ./arch/x86_64/kernel/machine_kexec.c
--- ../orig-linux-2.6.17/arch/x86_64/kernel/machine_kexec.c	2006-06-18 02:49:35.000000000 +0100
+++ ./arch/x86_64/kernel/machine_kexec.c	2007-01-08 15:01:25.000000000 +0000
@@ -149,8 +149,8 @@ typedef NORET_TYPE void (*relocate_new_k
 					unsigned long start_address,
 					unsigned long pgtable) ATTRIB_NORET;
 
-const extern unsigned char relocate_new_kernel[];
-const extern unsigned long relocate_new_kernel_size;
+extern const unsigned char relocate_new_kernel[];
+extern const unsigned long relocate_new_kernel_size;
 
 int machine_kexec_prepare(struct kimage *image)
 {
