diff -pruN ../orig-linux-2.6.17/include/linux/kernel.h ./include/linux/kernel.h
--- ../orig-linux-2.6.17/include/linux/kernel.h	2006-06-18 02:49:35.000000000 +0100
+++ ./include/linux/kernel.h	2007-01-08 15:38:20.000000000 +0000
@@ -114,6 +114,8 @@ extern int scnprintf(char * buf, size_t 
 	__attribute__ ((format (printf, 3, 4)));
 extern int vscnprintf(char *buf, size_t size, const char *fmt, va_list args)
 	__attribute__ ((format (printf, 3, 0)));
+extern char *kasprintf(gfp_t gfp, const char *fmt, ...)
+	__attribute__ ((format (printf, 2, 3)));
 
 extern int sscanf(const char *, const char *, ...)
 	__attribute__ ((format (scanf, 2, 3)));
diff -pruN ../orig-linux-2.6.17/lib/vsprintf.c ./lib/vsprintf.c
--- ../orig-linux-2.6.17/lib/vsprintf.c	2007-01-08 15:38:07.000000000 +0000
+++ ./lib/vsprintf.c	2007-01-08 15:38:20.000000000 +0000
@@ -849,3 +849,26 @@ int sscanf(const char * buf, const char 
 }
 
 EXPORT_SYMBOL(sscanf);
+
+
+/* Simplified asprintf. */
+char *kasprintf(gfp_t gfp, const char *fmt, ...)
+{
+	va_list ap;
+	unsigned int len;
+	char *p;
+
+	va_start(ap, fmt);
+	len = vsnprintf(NULL, 0, fmt, ap);
+	va_end(ap);
+
+	p = kmalloc(len+1, gfp);
+	if (!p)
+		return NULL;
+	va_start(ap, fmt);
+	vsnprintf(p, len+1, fmt, ap);
+	va_end(ap);
+	return p;
+}
+
+EXPORT_SYMBOL(kasprintf);
